src/Rector/ClassMethod/ResponseReturnTypeControllerActionRector.php

```diff
@@ -98,27 +98,98 @@ public function refactor(Node $node): ?Node
             return null;
         }
 
-        // has redirect return response
-        if ($this->hasRedirectReturnResponse($node)) {
+        if ($this->isRedirectResponseReturn($node)) {
             $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\RedirectResponse');
-        } else {
-            $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\Response');
+
+            return $node;
+        }
+        if ($this->isBinaryFileResponseReturn($node)) {
+            $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\BinaryFileResponse');
+
+            return $node;
+        }
+        if ($this->isJsonResponseReturn($node)) {
+            $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\JsonResponse');
+
+            return $node;
+        }
+        if ($this->isStreamedResponseReturn($node)) {
+            $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\StreamedResponse');
+
+            return $node;
         }
 
+        $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\Response');
+
         return $node;
     }
 
-    private function hasRedirectReturnResponse(ClassMethod $classMethod): bool
+    private function isRedirectResponseReturn(ClassMethod $classMethod): bool
+    {
+        $returns = $this->betterNodeFinder->findInstancesOfInFunctionLikeScoped($classMethod, Return_::class);
+
+        foreach ($returns as $return) {
+            if (! $return->expr instanceof MethodCall) {
+                return false;
+            }
+
+            $methodCall = $return->expr;
+            $functionName = $this->getName($methodCall->name);
+            if (! in_array($functionName, ['redirectToRoute', 'redirect'], true)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private function isBinaryFileResponseReturn(ClassMethod $classMethod): bool
+    {
+        $returns = $this->betterNodeFinder->findInstancesOfInFunctionLikeScoped($classMethod, Return_::class);
+
+        foreach ($returns as $return) {
+            if (! $return->expr instanceof MethodCall) {
+                return false;
+            }
+
+            $methodCall = $return->expr;
```

Comment:
@JohJohan it would be better to ensure check `$methodCall->var` is `Variable` named `this` so it will always from current class/from parent for next improvement 

```php
if (! $methodCall->var instanceof Variable || $methodCall->var->name !== 'this') {
```

Reply:
See https://github.com/rectorphp/rector-symfony/pull/309